bundles/org.openhab.binding.siemenshvac/src/main/java/org/openhab/binding/siemenshvac/internal/discovery/SiemensHvacDeviceDiscoveryService.java

```diff
@@ -0,0 +1,175 @@
+public class SiemensHvacDeviceDiscoveryService extends AbstractDiscoveryService
+        implements DiscoveryService, ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(SiemensHvacDeviceDiscoveryService.class);
+
+    private @Nullable SiemensHvacMetadataRegistry metadataRegistry;
+    private @Nullable SiemensHvacBridgeBaseThingHandler siemensHvacBridgeHandler;
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Set.of(SiemensHvacBindingConstants.THING_TYPE_OZW672);
+
+    private static final int SEARCH_TIME = 10;
+
+    public SiemensHvacDeviceDiscoveryService() {
+        super(SUPPORTED_THING_TYPES, SEARCH_TIME);
+    }
+
+    @Reference
+    public void setSiemensHvacMetadataRegistry(@Nullable SiemensHvacMetadataRegistry metadataRegistry) {
+        this.metadataRegistry = metadataRegistry;
+    }
+
+    public void unsetSiemensHvacMetadataRegistry(SiemensHvacMetadataRegistry metadataRegistry) {
+        this.metadataRegistry = null;
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+    }
+
+    private @Nullable ThingUID getThingUID(ThingTypeUID thingTypeUID, String serial) {
+        final SiemensHvacBridgeBaseThingHandler lcSiemensHvacBridgeHandler = siemensHvacBridgeHandler;
+        if (lcSiemensHvacBridgeHandler != null) {
+            ThingUID localBridgeUID = lcSiemensHvacBridgeHandler.getThing().getUID();
+            return new ThingUID(thingTypeUID, localBridgeUID, serial);
+        }
+        return null;
+    }
+
+    @Override
+    public void startScan() {
+        final SiemensHvacMetadataRegistry lcMetadataRegistry = metadataRegistry;
+        final SiemensHvacBridgeBaseThingHandler lcSiemensHvacBridgeHandler = siemensHvacBridgeHandler;
+        logger.debug("call startScan()");
+
+        if (lcMetadataRegistry != null) {
+            try {
+                lcMetadataRegistry.readMeta();
+            } catch (SiemensHvacException ex) {
+                logger.debug("Exception occurred during execution: {}", ex.getMessage(), ex);
+                return;
+            }
+
+            ArrayList<SiemensHvacMetadataDevice> devices = lcMetadataRegistry.getDevices();
+
+            if (devices == null) {
+                return;
+            }
+
+            for (SiemensHvacMetadataDevice device : devices) {
+
+                String name = device.getName();
+                String type = device.getType();
+                String addr = device.getAddr();
+                String serialNr = device.getSerialNr();
+
+                if (name.indexOf("OZW672") >= 0) {
+                    continue;
+                }
+
+                logger.debug("Find devices: {} / {} / {} / {}", name, type, addr, serialNr);
+
+                String typeSn = UidUtils.sanetizeId(type);
+                ThingTypeUID thingTypeUID = new ThingTypeUID(SiemensHvacBindingConstants.BINDING_ID, typeSn);
```

Comment:
I don't think it's a good thing to remove it.
In complex setup, you could have multiple slave devices (of the same type) as said in previous comment.
So serial is important so we don't intermix devices.

Reply:
I did not mean to drop serial number, only to potentially drop **type** as Thing ID if not strictly needed when you have the needed Thing configuration parameters?

Reply:
Oh, ok. 
I will check it after lunch.
I commit a bunch of fixes this morning in regards to your last review.
Will verify this afternoon if I miss something.

Reply:
Jacob,

Sorry, but I’ve definitively not understand what you want to said on this one.
I've got the following result in these methods:

ThingTypeUID are:
    siemenshvac:ozw672_01
    siemenshvac:rvs41_813_327

and thingUID are : 
siemenshvac:ozw672_01:5c66ebd975:00FD00FF45F4
siemenshvac:rvs41_813_327:5c66ebd975:00770000756A

where 5c66ebd975 is the bridgeId, and last part the serial of the device.*
Can you elaborate a little more on what you want me to modify ?

Reply:
As an example, in the Hue binding for API v2, there is only one Thing Type (besides the Bridge): "device". So I have a Thing UID etc. like this:
```yaml
UID: hue:device:home:traadfri_white_bulb4
label: Le Klint Snowdrop
thingTypeUID: hue:device
configuration:
  resourceId: 00000000-0000-0000-0000-000000000001
bridgeUID: hue:bridge-api2:home```
```

In API v1 there were many statically defined Thing Types, i.e. for each specific type of device (dimmable light, on/off light, motion sensor, etc.)

So the difference here is mostly static vs. dynamic.

In your case the binding dynamically creates everything, so the part I'm not getting is why you need Thing ID "rvs41_813_327" rather than simply "device" (as example)?

I would assume the serial number to be configured as a parameter, since it's needed to identify the device, like resourceId in my Hue example above. The user can manually create a Bridge and Thing and assign any name, so you could for example have `siemenshvac:rvs41_813_327:mybridge:mything`. Or with my suggestion: `siemenshvac:device:mybridge:mything`.

Reply:
Hello @jlaur,

I still miss something.
I give a try tonight try to modify the binding to align to your recommendation.

My interrogation is about the ThingType UID, Thing Uid composition:

    In the hue sample, you will have: hue:device:home:traadfri_white_bulb4
    In the current siemenshvac, something like: siemenshvac:rvsxxxx or siemenshav:ozwxxx or other.

As my understand, the first two segment is the thingTypesUID.
In the siemenshvacbinding, rvsxxx, or ozwxxxx are different thing type, with different channel sets.
So if I change to a generic "siemenshvac:device", I don't see how I can register my channel set to the different ThingsTypes.

Let me know if it makes senses, of If I miss something.

Reply:
My last uid scheme is as following:

![image](https://github.com/openhab/openhab-addons/assets/4678815/1fa7f1be-2d66-4b00-881b-1992afe17a54)

I think this misunderstanding come from the thingTypeUid taht need to be different because of differrent channel sets in regards to thingsTypeUid.

Im simplify the scheme to remove the SerialNumber that is not really needs.

Reply:
We are getting closer. I think one of the misunderstands is the difference between:
- Thing Type ID
- Thing Type UID
- Thing ID
- Thing UID

For the bridge, The Thing Type ID is now `ozw` and the Thing Type UID is `siemenshvac:ozw` - which is fine.

For the Thing here, I suggested something generic like `siemenshvac:device` as Thing Type UID rather than using the type, e.g. `siemenshvac:rvs41-813-327`. The reason for this is that you are creating all channel types and channels dynamically anyway, so you actually don't need different Thing Types.

When having statically declared Thing Types in XML, you will often create different Thing Types to represent difference kind of devices etc. For example, in the deCONZ binding you have these Thing Types:
https://www.openhab.org/addons/bindings/deconz/#supported-things

In this case, the Thing Type UID would be something like `deconz:powersensor`, which has a specific set of channels related to power measurement.

I realize now from the README that `rvs41-813-327` is something concrete, so it's not wrong, just a bit complex IMHO. In this case I would then propose to document `rvs41-813-327` and other supported types.

Reply:
Hello Jacob,

There is still something I've don't understand there.

If you put the same thing type UID for two different thing type, how can you have different channel definitions for each one.

I give it a try this week. I my installation, beside the gateway, I have two different thing device : the rvs41-813-327 wich is part of my PAC, and the ozw672-01 which is a virtual device implements by the gateway.

I try to give them same thing type as advice "siemenshvac:device", but when doing this I find that the rvs41-813-327 thing presents the channel set of the ozw672-01 things. 

I was thinking that the Thing Type UID was there to enable differenciating channel definition from one device from anoter. But perharps I miss something there.

Laurent.

Reply:
One more comment also about "In this case I would then propose to document rvs41-813-327 and other supported types."

I don't know the full set of supported devices. Perhaps there will be 10, or 50 devices, or even more.
But it shoud work out of the box because all devices use the same metadata discovery definitions.
So the binding won't have to add specific code for each supporting devices.
This is also respond to one of your questions in open comments about supportThings function.

Laurent.

Reply:
> If you put the same thing type UID for two different thing type, how can you have different channel definitions for each one.

I guess I managed to confuse myself about what's actually dynamically created in this binding. I'm not very experienced with `ChannelTypeRegistry`. I somehow wrongly assumed that the Thing handler was creating the channels dynamically, but I see now that it's not. My visits are apparently not frequent enough to remember how this binding works, or I may never have fully understood.

To your question: If the handler would dynamically create the appropriate channels for the given device, that would be the answer. However, it doesn't seem to be the case here.

So perhaps you can help me out a bit here, and sorry if we are running in circles: Can you explain how e.g. the `rvs41-813-327` is created and how its channels are created?

Maybe everything is fine, I just need to wrap my head around this.

Can you also show a screenshot of a `rvs41-813-327` being created manually, i.e. which configuration parameters are seen etc.?

Reply:
Let me ask differently: How can I manually create an `rvs41-813-327` Thing? Can I do this in a .things file without discovery?

![image](https://github.com/openhab/openhab-addons/assets/19519842/2dfa7b0e-6dbb-4524-92e1-a44071545dc8)

Reply:
Sub devices like rvs41 are registered to the gateway ozw672.

When I first initialized the gateway ozw672, I call some api endpoint to discover wich device are registered, and what are the structure of the device, and what are the channel and channel type that are supported. This discovery part enable me to construct the siemens.json.

After this, I create a ThingTypeUID for each device, and registered the corresponding channel in the ChannelTypeRegistry. 

This is far before the Thing Handler is even instantiate. So Channel creation is not dinamically create in the Thing Handler, but far before in the SiemensHvacThingTypeProvider / SiemensHvacChannelTypeProvider.

When you create the thing rvs41, you already have the channel definition list inside the thing, with associate ChannelType. So you only have to link the channel to your item.

Reply:
Thanks for the explanation. Would it be possible for you to attach an example of such `siemens.json` file?

Reply:
Yes, you can do it manualy also from the thing definition.
It was the reason of the samples inside the Readme.md.
It would be more complex of course.

For this, you will need to add a channel definition (what we removed from Readme.md)
String              Boiler_State_Pump_HWS       "HWS Pump State [%s]"                   { channel = "siemenshvac:RVS41_813_327:local:local:2237#2259_PpeChargeECS"              }  

The important part is the channel part : first is the ThingUid definition.
The seconde part : 2237#2259 is the id of the channel registered in the ozw672 gateway.
If you do it manually, you will have to check the correct value in the ozw672 web interface, and report it in the channel definition.

Reply:
Yes, of course.
Not sure if it is the more recent one, because i change of computer from last week from today.
[Uploading siemens.json…]()

Reply:
I'm currently switch to work on another binding, but can setup back my computer to the siemenshvac branch tomorrow.
If it can help, we can setup a remote session so I can show you more in details how it works.